### PR TITLE
[c10d] Make nccl2 the default NCCL backend

### PR DESCRIPTION
**Pull request by d4l3k**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #193241
* __->__ #192281
* #195485

The standard `nccl` backend still selected legacy ProcessGroupNCCL unless users opted into nccl2 through an environment variable, even though the explicit `nccl2` and `nccl-legacy` names make backend selection reversible. Invert the registration so an unset variable or `TORCH_DIST_USE_NCCL2=1` selects nccl2, while `TORCH_DIST_USE_NCCL2=0` and the explicit `nccl-legacy` name remain rollback paths.

Defaulting also exposed three nccl2 compatibility gaps: initialized CUDA processes reused device 0 across ranks, unsupported bitwise reductions raised `RuntimeError` instead of `ValueError`, and timeout text differed from legacy NCCL. Select fallback devices from global rank and match the legacy exception contracts.

> AI-assisted TorchTitan release-test status (2026-08-11). Except for the parity comparison, all runs used 4 GPUs with `TORCH_DIST_USE_NCCL2` unset, so `nccl` resolved to nccl2.
>
> | Area | Test | Matrix coverage | Status |
> | --- | --- | --- | --- |
> | Deterministic parity | Legacy NCCL vs default nccl2 | 10 training steps; `global_avg_loss`, `global_max_loss`, and `grad_norm` matched exactly | PASS |
> | TP + FSDP | `2d_eager_spmd_types` | 2 loss-function variants | PASS |
> | PP + FSDP | `pp_dp_1f1b_spmd_types` | 2 1F1B variants, including explicit layers per stage; exercises pipeline P2P | PASS |
> | DDP | `ddp_spmd_types` | 4-way replication | PASS |
> | HSDP | `hsdp_spmd_types` | 2-way replication x 2-way sharding | PASS |
> | CP | `cp_spmd_types` | 4-way context parallelism | PASS |
> | FSDP + CP | `fsdp+cp_spmd_types` | 2-way sharding x 2-way context parallelism | PASS |
> | DeepSeek V3 | `deepseek_v3_hsdp+ep_spmd_types` | HSDP + expert parallelism | PASS |
> | Qwen3 MoE | `qwen3_moe_fsdp+tp+ep_param_groups_spmd_types` | FSDP + tensor parallelism + expert parallelism with parameter groups | PASS |

Test plan:

```bash
BUILD_TEST=0 USE_FBGEMM=0 USE_NNPACK=0 USE_QNNPACK=0 USE_XNNPACK=0 USE_FLASH_ATTENTION=0 USE_MEM_EFF_ATTENTION=0 USE_DISTRIBUTED=1 USE_CUDA=1 VIRTUAL_ENV=/home/tristanr/pytorch/.venv PATH=/home/tristanr/pytorch/.venv/bin:$PATH /home/tristanr/.local/bin/uv pip install -e . -v --no-build-isolation
```

```bash
VIRTUAL_ENV=/home/tristanr/pytorch/.venv /home/tristanr/.local/bin/uv run --no-project python test/distributed/test_c10d_common.py -k nccl_backend_registration
```

```bash
TORCH_DIST_USE_NCCL2=0 PATH=/home/tristanr/pytorch/.venv/bin:$PATH VIRTUAL_ENV=/home/tristanr/pytorch/.venv NGPU=4 LOG_RANK=0 ./run_train.sh --dump_folder=/home/tristanr/pytorch/agent_space/nccl2-default-release/legacy_job_v3 --debug.deterministic --debug.seed=42 --metrics.enable_tensorboard --metrics.log_freq=1 --training.steps=10 --metrics.save_tb_folder=tb_legacy
```

```bash
PATH=/home/tristanr/pytorch/.venv/bin:$PATH VIRTUAL_ENV=/home/tristanr/pytorch/.venv NGPU=4 LOG_RANK=0 ./run_train.sh --dump_folder=/home/tristanr/pytorch/agent_space/nccl2-default-release/nccl2_job --debug.deterministic --debug.seed=42 --metrics.enable_tensorboard --metrics.log_freq=1 --training.steps=10 --metrics.save_tb_folder=tb_default
```

The 10-step TensorBoard series for `loss_metrics/global_avg_loss`, `loss_metrics/global_max_loss`, and `grad_norm` were identical between the two deterministic runs.

```bash
set -euo pipefail
unset TORCH_DIST_USE_NCCL2
export PATH=/home/tristanr/pytorch/.venv/bin:$PATH
export VIRTUAL_ENV=/home/tristanr/pytorch/.venv
for test_name in 2d_eager_spmd_types pp_dp_1f1b_spmd_types ddp_spmd_types hsdp_spmd_types cp_spmd_types fsdp+cp_spmd_types; do
  /home/tristanr/pytorch/.venv/bin/python -m tests.integration_tests.run_tests "/home/tristanr/pytorch/agent_space/nccl2-default-release/features_${test_name}" --test_suite features --test_name "$test_name" --ngpu 4 --no-parallel
done
for test_name in deepseek_v3_hsdp+ep_spmd_types qwen3_moe_fsdp+tp+ep_param_groups_spmd_types; do
  /home/tristanr/pytorch/.venv/bin/python -m tests.integration_tests.run_tests "/home/tristanr/pytorch/agent_space/nccl2-default-release/models_${test_name}" --test_suite models --test_name "$test_name" --ngpu 4 --no-parallel
done
```

```bash
.venv/bin/spin quicklint
env PATH=/home/tristanr/pytorch/.venv/bin:/home/tristanr/.local/bin:/usr/local/bin:/usr/bin:/bin .venv/bin/lintrunner -a
```

```bash
env VIRTUAL_ENV=/home/tristanr/pytorch/.venv PATH=/home/tristanr/pytorch/.venv/bin:/home/tristanr/.local/bin:/usr/local/bin:/usr/bin:/bin /home/tristanr/.local/bin/uv run --no-project test/distributed/test_c10d_common.py BackendEntryPointTest.test_nccl2_device_uses_rank_without_local_rank
```

```bash
env CUDA_VISIBLE_DEVICES=0,1 NCCL_DEBUG=WARN VIRTUAL_ENV=/home/tristanr/pytorch/.venv PATH=/home/tristanr/pytorch/.venv/bin:/home/tristanr/.local/bin:/usr/local/bin:/usr/bin:/bin /home/tristanr/.local/bin/uv run --no-project test/distributed/test_c10d_ops_nccl.py ProcessGroupNCCLOpTest.test_allreduce_ops
```

```bash
env CUDA_VISIBLE_DEVICES=0,1 NCCL_DEBUG=WARN VIRTUAL_ENV=/home/tristanr/pytorch/.venv PATH=/home/tristanr/pytorch/.venv/bin:/home/tristanr/.local/bin:/usr/local/bin:/usr/bin:/bin /home/tristanr/.local/bin/uv run --no-project test/distributed/test_c10d_nccl2.py ProcessGroupNCCL2BlockingWaitTest.test_wait_times_out_and_process_survives
```

```bash
env CUDA_VISIBLE_DEVICES=0,1 NCCL_DEBUG=WARN VIRTUAL_ENV=/home/tristanr/pytorch/.venv PATH=/home/tristanr/pytorch/.venv/bin:/home/tristanr/.local/bin:/usr/local/bin:/usr/bin:/bin /home/tristanr/.local/bin/uv run --no-project test/distributed/test_c10d_nccl.py ProcessGroupNCCLGroupTest.test_extend_nccl_pg_timeout_backend0
```

Authored with an AI assistant.

### CODE REVIEW

**Comment by kapilsh** [comment]
🚀 🚀 🚀

**Review by kapilsh — APPROVED** [review]
Accepting so you can land when all required changes are landed.

**Review by tushar00jain — APPROVED** [review]

**Review by kapilsh — APPROVED** [review]

**Review by dolpm — APPROVED** [review]
!!!!
